### PR TITLE
NGRX for User info, Messages and Conversations #56

### PR DESCRIPTION
**Pull request by adrnchk**

### CODE REVIEW

File: DogeNetWebClient/src/app/actions/conversation.actions.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { Action } from '@ngrx/store';
+import { ConversationDetailsModel } from 'src/app/core/api/models';
+
+export const SET_CONVERSATIONS = '[Conversations] Set';
+export const ADD_CONVERSATION = '[Conversations] Add';
+export const DELETE_CONVERSATION = '[Conversations] Delete';
+export const CLEAR_CONVERSATIONS = '[Conversations] Clear';
+
+export class SetConversations implements Action {
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/actions/conversation.actions.ts]: This is an old approach to creating actions, please use new one (https://ngrx.io/guide/store/actions)

File: DogeNetWebClient/src/app/actions/messages.action.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+export const DELETE_MESSAGE = '[Messages] Delete';
+export const EDIT_MESSAGE = '[Messages] Edit';
+
+export class SetNewDialogue implements Action {
+  readonly type = SET_NEW_DIALOGUE;
+  constructor(
+    public payload: { id: number; messages: MessagesDetailsModel[] }
+  ) {}
+}
+export class SendMessage implements Action {
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/actions/messages.action.ts]: Same here

File: DogeNetWebClient/src/app/reducers/messages.reducer.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+export const messagesNode = 'messages';
+
+export const messagesReducer = (
+  state = initialState,
+  action: Actions.All
+): MessagesState => {
+  switch (action.type) {
+    case Actions.SET_NEW_DIALOGUE:
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/reducers/messages.reducer.ts]: Use new approach for creating reducers here

File: DogeNetWebClient/src/app/reducers/index.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  [messagesNode]: MessagesState;
+}
+
+export const reducers: ActionReducerMap<State, any> = {
+  [userNode]: userReducer,
+  [conversationsNode]: conversationsReducer,
+  [messagesNode]: messagesReducer,
+};
+
+export const metaReducers: MetaReducer<State, any>[] = !environment.production
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/reducers/index.ts]: What is this?

File: DogeNetWebClient/src/app/reducers/user-info.reducer.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { Action } from '@ngrx/store';
+import * as UserActions from 'src/app/actions/user-info.actions';
+import { AccountDetailsModel } from 'src/app/core/api/models';
+
+export interface UserState {
+  info: AccountDetailsModel;
+}
+
+const initialState: UserState = {
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/reducers/user-info.reducer.ts]: Why state is here but not in separate?

File: DogeNetWebClient/src/app/reducers/index.ts (hunk 4 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/reducers/index.ts]: @oleh-leobit, this is something like middleware that works before reducer

File: DogeNetWebClient/src/app/actions/conversation.actions.ts (hunk 1 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/actions/conversation.actions.ts]: done

File: DogeNetWebClient/src/app/actions/messages.action.ts (hunk 2 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/actions/messages.action.ts]: done

File: DogeNetWebClient/src/app/reducers/messages.reducer.ts (hunk 3 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/reducers/messages.reducer.ts]: done

File: DogeNetWebClient/src/app/reducers/user-info.reducer.ts (hunk 5 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/reducers/user-info.reducer.ts]: done

File: DogeNetWebClient/src/app/store/actions/conversation.actions.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { createAction, props } from '@ngrx/store';
+import { ConversationDetailsModel } from 'src/app/core/api/models';
+
+export const SET_CONVERSATIONS = '[Conversations] Set';
+export const ADD_CONVERSATION = '[Conversations] Add';
+export const DELETE_CONVERSATION = '[Conversations] Delete';
+export const CLEAR_CONVERSATIONS = '[Conversations] Clear';
+
+export const SetConversations = createAction(
+  SET_CONVERSATIONS,
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/store/actions/conversation.actions.ts]: Do we need a separate variable for the action name?

File: DogeNetWebClient/src/app/store/reducers/conversations.reducer.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import * as Actions from 'src/app/store/actions/conversation.actions';
+import { ConversationDetailsModel } from 'src/app/core/api/models';
+import { ConversationsState } from '../states/ConversationsState';
+import { createReducer, on } from '@ngrx/store';
+
+const initialState: ConversationsState = {
+  items: new Array<ConversationDetailsModel>(),
+};
+
+export const conversationsNode = 'conversations';
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/store/reducers/conversations.reducer.ts:5]: What is this string for and why it is placed here?

File: DogeNetWebClient/src/app/store/reducers/index.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  [messagesNode]: MessagesState;
+}
+
+export const reducers: ActionReducerMap<State, any> = {
+  [userNode]: userReducer,
+  [conversationsNode]: conversationsReducer,
+  [messagesNode]: messagesReducer,
+};
+
+export const metaReducers: MetaReducer<State, any>[] = !environment.production
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/store/reducers/index.ts]: I don't see any logic here. Why do we have to check the environment and return the same empty array? Seems these meta reducers should be removed from the code.

File: DogeNetWebClient/src/app/store/reducers/messages.reducer.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import * as Actions from 'src/app/store/actions/messages.action';
+import { Dialogue, MessagesState } from '../states/MessagesState';
+import { createReducer, on } from '@ngrx/store';
+
+const initialState: MessagesState = {
+  items: new Array<Dialogue>(),
+};
+
+export const messagesNode = 'messages';
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/store/reducers/messages.reducer.ts:5]: Same here

File: DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { Action, createReducer, on } from '@ngrx/store';
+import * as Actions from 'src/app/store/actions/user-info.actions';
+import { AccountDetailsModel } from 'src/app/core/api/models';
+import { UserState } from '../states/UserState';
+
+const initialState: UserState = {
```
Comment (oleh-leobit) [inline DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts]: Could we move the state into the separate file as it would be used separatly?

**Review by oleh-leobit — COMMENTED** [review]
Please, have a look on comments

File: DogeNetWebClient/src/app/store/actions/conversation.actions.ts (hunk 6 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/store/actions/conversation.actions.ts]: deleted

File: DogeNetWebClient/src/app/store/reducers/conversations.reducer.ts (hunk 7 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/store/reducers/conversations.reducer.ts:5]: @oleh-leobit ,we need to register the key under which the reducer will be stored. Then we create a {
image
}to determine what type of data should be returned by this editor. In order not to write constant labels in the index file and the selector is better to export them

File: DogeNetWebClient/src/app/store/reducers/index.ts (hunk 8 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/store/reducers/index.ts]: ok

File: DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts (hunk 10 above)
Comment (adrnchk) [reply DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts]: @oleh-leobit do you mean initialState?

Comment (adrnchk) [reply DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts]: one file for state interface one for reducer and one for initial state?

Comment (adrnchk) [reply DogeNetWebClient/src/app/store/reducers/user-info.reducer.ts]: or could I just export initial state from state file?

**Pull request merged**